New guide for the EditContext API (#32027)
mdn/content: files/en-us/web/api/editcontext_api/guide/index.md

- The DOM node that contains the start of the selection.
- A number representing the character position of the selection start, within the start node.
- The DOM node that contains the end of the selection.
- A number representing the character position of the selection end, within the end node.

```js
function render(text, selectionStart, selectionEnd) {
  // ...
  // The beginning of the render function is omitted for brevity.

  // Convert the start/end offsets to a DOM selection.
  const { anchorNode, anchorOffset, extentNode, extentOffset } =
    fromOffsetsToSelection(selectionStart, selectionEnd);

  // Render the selection in the editor element.
  document
    .getSelection()
    .setBaseAndExtent(anchorNode, anchorOffset, extentNode, extentOffset);
}
```

> You can see the code for the `fromOffsetsToSelection()` function in the [converter.js](<(https://github.com/mdn/dom-examples/tree/main/edit-context/html-editor/converter.js)>) file.

Reviewer: ```suggestion
You can see the code for the `fromOffsetsToSelection()` function in the [converter.js](<(https://github.com/mdn/dom-examples/tree/main/edit-context/html-editor/converter.js)>) file.
```

---
title: Using the EditContext API
slug: Web/API/EditContext_API/Guide
page-type: guide
---

{{DefaultAPISidebar("EditContext API")}}

The **[EditContext API](/en-US/docs/Web/API/EditContext_API)** can be used to build rich text editors on the web that support advanced text input experiences, such as {{glossary("Input Method Editor")}} (IME) composition, emoji picker, or any other platform-specific editing-related UI surfaces.

This article goes through all the steps that are necessary to build a text editor using the EditContext API. In this guide, you will build a simple HTML code editor that highlights the syntax of the code as you type, and that supports IME composition.

## Final code and live demo

If you want to see the final code, you can check out the [live demo](https://mdn.github.io/dom-examples/editcontext-api/html-editor/) and [source code](https://github.com/mdn/dom-examples/tree/master/editcontext-api/html-editor) on GitHub.

## Create the editor UI

> The first step is to create the UI for the editor. The editor will be a simple [`<div>`](/en-US/docs/Web/HTML/Element/div) element with he [`spellcheck`](/en-US/docs/Web/HTML/Global_attributes/spellcheck) attribute set to `false` to disable spell checking:

Reviewer 2: ```suggestion
The first step is to create the UI for the editor. The editor will be a simple {{HTMLElement("div")}} element with the [`spellcheck`](/en-US/docs/Web/HTML/Global_attributes/spellcheck) attribute set to `false` to disable spell checking:
```

```html
<div id="html-editor" spellcheck="false"></div>
```

> The following CSS code is used to make the editor fill the entire viewport and scroll when there's too much content to fit. The [`white-space`](/en-US/docs/Web/CSS/white-space) property is also used to preserve whitespace, and the [`tab-size`](/en-US/docs/Web/CSS/tab-size) property is used to make tab characters render as two spaces. Finally, some default background and text colors are set.

Reviewer 2: ```suggestion
The following CSS code is used to make the editor fill the entire viewport and scroll when there's too much content to fit. The {{CSSXref("white-space")}} property is also used to preserve whitespace, and the {{CSSXref("tab-size")}} property is used to make tab characters render as two spaces. Finally, some default background and text colors are set.
```

```css
#html-editor {
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  padding: 1rem;
  overflow: auto;
  white-space: pre;
  tab-size: 2;
  background: #000;
  color: red;
}
```

## Make the editor editable

> To make an element editable on the web, most of the time, you use an [`<input>`](/en-US/docs/Web/HTML/Element/input) element, a [`<textarea>`](/en-US/docs/Web/HTML/Element/textarea) element, or the [`contenteditable`](/en-US/docs/Web/HTML/Global_attributes/contenteditable) attribute.

Reviewer 2: ```suggestion
To make an element editable on the web, most of the time, you use an {{HTMLElement("input")}} element, a {{HTMLElement("textarea")}} element, or the [`contenteditable`](/en-US/docs/Web/HTML/Global_attributes/contenteditable) attribute.
```

However, with the EditContext API, you can make other types of elements editable without using an attribute. To see the list of elements that can be used with the EditContext API, see [Possible elements](/en-US/docs/Web/API/HTMLElement/editContext#possible_elements) on the HTMLElement editContext property page.

> To make the editor editable, you create an {{domxref("EditContext")}} instance, and then you set the [`editContext`](/en-US/docs/Web/API/HTMLElement/editContext) property of the editor element to the `EditContext` instance:

Reviewer 2: ```suggestion
To make the editor editable, you create an {{domxref("EditContext")}} instance, and then you set the {{DOMxRef("HTMLElement.editContext", "editContext")}} property of the editor element to the `EditContext` instance:
```

To render the selection, because the EditContext API gives you the selection as a pair of start and end character offsets, you'll need to convert these offsets into a DOM selection. The following code snippet shows how you can do this:

```js
function convertFromOffsetsToSelection(start, end) {
  // Create a TreeWalker that walks over all the
  // text nodes in the editor element.
  const treeWalker = document.createTreeWalker(editorEl, NodeFilter.SHOW_TEXT);

  let offset = 0;
  let anchorNode = null;
  let anchorOffset = 0;
  let extentNode = null;
  let extentOffset = 0;

  // Iterate over the text nodes.
  while (treeWalker.nextNode()) {
    const node = treeWalker.currentNode;

    if (!anchorNode && offset + node.textContent.length >= start) {
      // The anchor node is the first node that contains the start offset.
      anchorNode = node;
      anchorOffset = start - offset;
    }

>     if (offset + node.textContent.length >= end) {

Reviewer 3: Note, the spec has been updated so that EditContext.updateSelection now allows backwards selection, and this change has been implemented in Chrome Canary, but won't hit Stable for a while yet: [5235427: [EditContext] Allow backwards selection](https://chromium-review.googlesource.com/c/chromium/src/+/5235427)

Looks like this doesn't necessarily handle backwards selection well. Do you think it should, or should we leave out that detail to keep things simple?

Author: I would prefer to make the change now, because it already allows backwards selections in Canary, and by the time people really start caring about the docs for this, I'm fairly sure the feature will be in Stable too.
So I'll update the demo repo, and the code here accordingly.

```js
function convertFromSelectionToOffsets(selection) {
  // Create a TreeWalker that walks over all the
  // text nodes in the editor element.
  const treeWalker = document.createTreeWalker(editorEl, NodeFilter.SHOW_TEXT);

  let anchorNodeFound = false;
  let extentNodeFound = false;
  let anchorOffset = 0;
  let extentOffset = 0;

  // Iterate over the text nodes until we find
  // the anchor and extent nodes given by the selection.
  while (treeWalker.nextNode()) {
    const node = treeWalker.currentNode;
    if (node === selection.anchorNode) {
      anchorNodeFound = true;
      anchorOffset += selection.anchorOffset;
    }

    if (node === selection.extentNode) {
      extentNodeFound = true;
      extentOffset += selection.extentOffset;
    }

    if (!anchorNodeFound) {
      anchorOffset += node.textContent.length;
    }
    if (!extentNodeFound) {
      extentOffset += node.textContent.length;
    }
  }

  // Return null if the nodes couldn't be found.
  // The selection is outside the editor element.
  if (!anchorNodeFound || !extentNodeFound) {
    return null;
  }

>   if (anchorOffset > extentOffset) {

Reviewer 3: Same comment as above about backwards selection.

One option would be to leave this as-is for now and then update the examples once https://chromium-review.googlesource.com/c/chromium/src/+/5235427 hits stable.

Author: I would prefer to make the change now, because it already allows backwards selections in Canary, and by the time people really start caring about the docs for this, I'm fairly sure the feature will be in Stable too.
So I'll update the demo repo, and the code here accordingly.

      (token) => token.pos <= offset && token.pos + token.value.length > offset,
    );
    if (token) {
      tokenNodes.push({
        node: token.node,
        nodeOffset: token.pos,
        charOffset: offset,
      });
    }
  }

  return tokenNodes;
}
```

With the above function, you can now listen to the `characterboundsupdate` event, and use it to calculate the bounds of the characters in the editor element:

```js
editContext.addEventListener("characterboundsupdate", (e) => {
  // Retrieve information about the token nodes in the range.
  const tokenNodes = convertFromOffsetsToListOfRenderedTokenNodes(
    e.rangeStart,
    e.rangeEnd,
  );

  // Convert this information into a list of DOMRect objects.
  const charBounds = tokenNodes.map(({ node, nodeOffset, charOffset }) => {
    const range = document.createRange();
    range.setStart(node.firstChild, charOffset - nodeOffset);
    range.setEnd(node.firstChild, charOffset - nodeOffset + 1);
    return range.getBoundingClientRect();
  });

  // Let the EditContext instance know about the character bounds.
  editContext.updateCharacterBounds(e.rangeStart, charBounds);
});
```

## Apply IME composition text formats

> For the final step, there is one more thing you need to do to fully support IME composition. When the user is composing text with an IME, the IME might decide that certain parts of the text being composed should be formatted differently to indicate the composition state. For example, the IME might decide to render the text in a different color, or to underline it.

Reviewer 3: Consider leaving out the mention of "color" here; `textformatupdate` requires the author to apply underlines of different styles or thickness but doesn't affect color.

As shown in the following code snippet, you will use the `convertFromOffsetsToSelection` function again to find the DOM nodes that are in the range of the IME composition. Let's also use the [CSS Custom Highlight API](/en-US/docs/Web/API/CSS_Custom_Highlight_API) to render the text formats. The CSS Custom Highlight API provides a mechanism for styling arbitrary text ranges by using JavaScript to create the ranges, and CSS to style them.

First, use the {{cssxref("::highlight", "::highlight()")}} pseudo-element to define the highlight styles. The text formats define the style and thickness of an underline. So, create as many highlights as there are possible combinations of underline styles and thicknesses:

```css
::highlight(ime-solid-thin) {
  text-decoration: underline 1px;
}

::highlight(ime-solid-thick) {
  text-decoration: underline 2px;
}

::highlight(ime-double-thin) {

>   text-decoration: underline double 1px;

Reviewer 3: No change required now, but just FYI `double` may be removed: https://github.com/w3c/edit-context/issues/89

  render();
});

// Do the initial render.
render();
```

### Styling the tokens

As seen in the previous `render()` function code example, each token is given a class name that corresponds to the type of token it is. The demo app uses this class name to style the tokens, using CSS, as shown below:

```css
.token-attributeName {
  color: rgb(207, 81, 198);
}

.token-attributeValue {
  font-style: italic;
  color: rgb(127 230 127);
  border: 1px dashed #8c8c8c;
  border-width: 1px 0 1px 0;
}
```

### Rendering the selection

Even though the demo app uses a `<div>` element for the editor, which already supports displaying a blinking text cursor and highlighting user selections, the EditContext API still requires us to render the selection. This is because the EditContext API can be used with other types of elements that don't support these behaviors. Rendering the selection ourselves also gives us more control over how the selection is displayed. Finally, because the `render()` function clears the HTML content of the editor element everytime it runs, any selection that the user might have made is lost the next time the `render()` function runs.

> To render the selection, the demo app uses the {{DOMxRef("Selection.setBaseAndExtent()")}} method at the end of the `render()` function. To use the `setBaseAndExtent()` method, we need a pair of DOM nodes and character offsets that represent the start and end of the selection. However, the EditContext API maintains the state for the current selection as a pair of start and end character offsets only. The demo app code uses another function, called `convertFromOffsetsToSelection()` that's used to convert these character offsets into a DOM selection.

Reviewer: I found this a little hard to understand and wonder if it is worth spelling out a bit.

```suggestion
To render the selection, the demo app uses the {{DOMxRef("Selection.setBaseAndExtent()")}} method at the end of the `render()` function. To use the `setBaseAndExtent()` method, we need a pair of DOM nodes and character offsets that represent the start and end of the selection. However, the EditContext API maintains the state for the current selection only as a pair of start and end character offsets into the entire edit buffer. The demo app code uses another function, called `convertFromOffsetsToSelection()` that's used to convert these character offsets into four values:
- the DOM node that contains the start of the selection
- a number representing the character position of the selection start, within the start node
- the DOM node that contains the end of the selection
- a number representing the character position of the selection end, within the end node
```

maybe??

Also as I said before, maybe worth putting this function and the corresponding one `convertFromSelectionToOffsets()` in a separate module. It might help emphasize that they are inverses.

```js
function render() {
  // ...
  // The beginning of the render function is omitted for brevity.

  // Render the selection at the correct location by
  // converting the start/end offsets to a DOM selection.
  const selectionStart = editContext.selectionStart;

>   const selectionEnd = editContext.selectionEnd;

Reviewer: In the code these are resolved as parameters. I'm not that clear why there is a divergence here.

The demo app uses the {{domxref("EditContext.updateControlBounds()")}} method, providing it with a {{domxref("DOMRect")}} object that represents the bounds of the editable text region. The demo app calls this method when the editor is initialized, and again when the window is resized:

```js
function updateControlBounds() {
  // Get the DOMRect object for the editor element.
  const editorBounds = editorEl.getBoundingClientRect();

  // Update the control bounds of the EditContext instance.
  editContext.updateControlBounds(editorBounds);
}

// Call the updateControlBounds function when the editor is initialized,
updateControlBounds();

// And call it again when the window is resized.
window.addEventListener("resize", updateControlBounds);
```

## Handling Tab, Enter, and other text edit keys

The `textupdate` event used in the previous section isn't fired when the user presses the <kbd>Tab</kbd> or <kbd>Enter</kbd> keys, so we need to handle these keys separately.

> To handle them, the demo app uses an event listener for the {{domxref("keydown_event", "keydown")}} event on the editor element, and uses this listener to update the `EditContext` instance's text content and selection, as shown below:

Reviewer: ```suggestion
To handle them, the demo app uses an event listener for the {{domxref("Element.keydown_event", "keydown")}} event on the editor element, and uses this listener to update the `EditContext` instance's text content and selection, as shown below:
```

However, because the EditContext API uses character offsets to represent the selection, the demo app also uses a function, `convertFromSelectionToOffsets()` that converts DOM selection objects to character offsets.

```js
// Listen to selectionchange events to let the
// EditContext know where it is.
document.addEventListener("selectionchange", () => {
  const selection = document.getSelection();

  // Convert the DOM selection into character offsets.
  const offsets = convertFromSelectionToOffsets(selection);
  if (!offsets) {
    return;
  }

  // Update the selection.
  editContext.updateSelection(offsets.start, offsets.end);

  // Get the bounds of the selection.
  editContext.updateSelectionBounds(
    selection.getRangeAt(0).getBoundingClientRect(),

>   );

Reviewer: This is different in the code (it's factored into an `updateSelection()` method).

As shown in the following code snippet, the demo app uses the `textformatupdate` event and the `convertFromOffsetsToSelection()` function again to find the text range that the IME composition wants to format.

```js
editContext.addEventListener("textformatupdate", (e) => {
  // Get the list of formats that the IME wants to apply.
  const formats = e.getTextFormats();

  // Iterate over the text formats.
  for (const format of formats) {
    // For each format, we know the underline style and thickness,
    // as well as the start and end offsets of the range that needs
    // to be styled.
    const { rangeStart, rangeEnd, underlineStyle, underlineThickness } = format;

    // Find the nodes that are in the range of the current format.
    const { anchorNode, anchorOffset, extentNode, extentOffset } =
      convertFromOffsetsToSelection(rangeStart, rangeEnd);

>     // Style the above range...

Reviewer: I'm not sure if this is worth doing, but...I understand that you want to show things in a particular order here:

1) getting the ranges and formats, in the handler
2) digressing a bit into the highlights API, setting up the highlights
3) back in the handler, applying the highlights

But it is a little odd to repeat the handler, first without (1) and then with (3) with the highlighting code. Perhaps to avoid this you could factor the code that applies the highlights:

```js
          const highlight = imeHighlights[`${format.underlineStyle.toLowerCase()}-${format.underlineThickness.toLowerCase()}`];
          if (highlight) {
            const range = document.createRange();
            range.setStart(anchorNode, anchorOffset);
            range.setEnd(extentNode, extentOffset);
            highlight.add(range);
          }
```

..into a separate function, like `addHighlights()` or something. Then you could first show the complete handler, calling the function, then digress into the highlights API, then show the `addHighlights()`.

But like I say, not sure it is worth it.

  while (treeWalker.nextNode()) {
    const node = treeWalker.currentNode;

    if (!anchorNode && offset + node.textContent.length >= orderedStart) {
      // The anchor node is the first node that contains the start offset.
      anchorNode = node;
      anchorOffset = orderedStart - offset;
    }

>     if (offset + node.textContent.length >= orderedEnd) {

Reviewer 3: Could we eliminate the reversal at the beginning and end of this function by changing this to the following (so that it works the same as the search for the anchorNode, and the loop can keep going if the extentNode is found first)?

    if (!extendNote && offset + node.textContent.length >= orderedEnd) {
      // The extent node is the first node that contains the end offset.
      extentNode = node;
      extentOffset = orderedEnd - offset;
    }

And then optionally:
```
if (anchorNode && extentNode) {
  break;
}
```

Author: Done in https://github.com/mdn/dom-examples/pull/252, thanks!
